lib/drivers/bayerContourNext.js

```diff
@@ -0,0 +1,510 @@
+/*
+ * == BSD2 LICENSE ==
+ * Copyright (c) 2014, Tidepool Project
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the associated License, which is identical to the BSD 2-Clause
+ * License as published by the Open Source Initiative at opensource.org.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the License for more details.
+ *
+ * You should have received a copy of the License along with this program; if
+ * not, you can obtain one from Tidepool Project at tidepool.org.
+ * == BSD2 LICENSE ==
+ */
+
+/**
+ * IOET's code start here
+ * packet frame format <STX> FN text <ETX> C1 C2 <CR> <LF>
+ * */
+var _ = require('lodash');
+var async = require('async');
+var sundial = require('sundial');
+var crcCalculator = require('../crc.js');
+var struct = require('../struct.js')();
+var annotate = require('../eventAnnotations');
+
+var isBrowser = typeof window !== 'undefined';
+var debug = isBrowser ? require('../bows')('BCNextDriver') : debug;
+
+
+module.exports = function (config) {
+  var cfg = _.clone(config);
+  var hidDevice = config.deviceComms;
+
+  var STX = 0x02;
+  var ETB = 0x17;
+  var ETX = 0x03;
+  var EOT = 0x04;
+  var HID_PACKET_SIZE = 60;
+
+  /*
+   * TODO This info should be collected in the initial phase cause we can't assert if
+   *      it is the same for all contour devices
+   * */
+  var REPORT_BYTES = {
+    reportID: 0x41, // A
+    checksum: 0x00,
+    hostID  : 0x42, // B
+    deviceID: 0x43  // C
+  };
+
+  /* end */
+
+  var ASCII_CONTROL = {
+    ACK : 0x06,
+    NAK : 0x15,
+    ENQ : 0x05
+  };
+
+  var probe = function(cb){
+    debug('attempting probe of Bayer Contour Next');
+  };
+
+  var bcnPacketHandler = function (buffer) {
+    // we need to get the REPORT_BYTES here for the first communication
+    // time if they are not set
+    var discardCount = 3;
+    if (discardCount) {
+      buffer.discard(discardCount);
+    }
+
+    if (buffer.len() < 6) { // all complete packets must be at least this long
+      return false;       // not enough there yet
+    }
+
+    // there's enough there to try, anyway
+    var packet = extractPacket(buffer.bytes());
+    if (packet.packet_len !== 0) {
+      // remove the now-processed packet
+      buffer.discard(HID_PACKET_SIZE + 1);  // 1 is for packet_length
+    }
+
+    if (packet.valid) {
+      return packet;
+    } else {
+      return null;
+    }
+  };
+
+  var buildPacket = function (command, cmdlength) {
+    var datalen = cmdlength + 4; // We need just 5 bytes for now
+    var buf = new ArrayBuffer(datalen);
+    var bytes = new Uint8Array(buf);
+    var ctr = struct.pack(bytes, 0, 'bbb', REPORT_BYTES.reportID,
```

Comment:
we need to add ABC always before sending anything, so this constructor only adds those bytes